### PR TITLE
improve(disputer): enable arbitrary decimal combinations between collateral, synthetic and price feed in disputer bot

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

issue #2351 has a fair amount of detail on the problem space.

Follows very closely on from https://github.com/UMAprotocol/protocol/pull/2276 

**Summary**

The disputer re-uses the emp client in a very similar way to how the liquidator uses it. As a result, this PR simply updates the disputer to correctly implement the new client interfaces and updates the tests accordingly. Please see #2351 and https://github.com/UMAprotocol/protocol/pull/2276  for details on the original change made to the client.

**Issue(s)**

issue #2351

### CODE REVIEW

File: packages/common/src/FormattingUtils.js [hunk 1]
```diff
@@ -124,9 +124,9 @@ function addSign(number) {
   }
 }
 
-// Take an amount based on fromDecimals and convert it to an amount based on toDecimals
-// For example amount = 100 usdt, its decimals are 6. You want to convert it to 18.
-// convertDecimals(6,18)(100000000)  => 1000000000000000000000.
+// Take an amount based on fromDecimals and convert it to an amount based on toDecimals. For example 100 usdt = 100e6,
+// with 6 decimals. If you wanted to convert this to a base 18 decimals you would get:
```
Comment (nicholaspai) [inline packages/common/src/FormattingUtils.js:128]: It looks like this function catches both cases, where `fromDecimals > toDecimals` and `fromDecimals < toDecimals`, right?

File: packages/disputer/index.js [hunk 2]
```diff
@@ -99,13 +98,22 @@ async function run({
...
     });
 
     // Client and dispute bot.
-    const empClient = new ExpiringMultiPartyClient(logger, getAbi("ExpiringMultiParty"), web3, empAddress);
+    const empClient = new ExpiringMultiPartyClient(
+      logger,
+      getAbi("ExpiringMultiParty"),
+      web3,
+      empAddress,
+      collateralTokenDecimals,
```
Comment (nicholaspai) [inline packages/disputer/index.js:112]: Shouldn't the `collateralTokenDecimals == syntheticTokenDecimals` for all future  EMP's, if so then why pass both of them in here

File: packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js [hunk 3]
```diff
@@ -49,6 +49,7 @@ const defaultConfigs = {
     lookback: 7200,
     invertPrice: true,
     minTimeBetweenUpdates: 60,
+    priceFeedDecimals: 8,
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js:52]: Should we also add this exact same override to the BCHNBTC pricefeed?

File: packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js [hunk 4]
```diff
@@ -72,6 +72,17 @@ class MedianizerPriceFeed extends PriceFeedInterface {
     return Math.max(...lastUpdateTimes);
   }
 
+  // Gets the decimals of the medianized price feeds. Errors out if any price feed had a different number of decimals.
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/price-feed/MedianizerPriceFeed.js:75]: Good check

File: packages/liquidator/index.js [hunk 5]
```diff
@@ -174,6 +184,15 @@ async function run({
       liquidatorConfig
     });
 
+    logger.debug({
+      at: "Liquidator#index",
+      message: "Liquidator initialized",
+      collateralDecimals: Number(collateralTokenDecimals),
+      syntheticDecimals: Number(syntheticTokenDecimals),
+      priceFeedDecimals: Number(priceFeed.getPriceFeedDecimals()),
```
Comment (nicholaspai) [inline packages/liquidator/index.js:192]: Is this the only case where a pricefeed's `getPriceFeedDecimals()` is used? If so, how do you anticipate this method can be used.

**Review by nicholaspai — COMMENTED** [review]
Great start overall, have a question about the new `getPriceFeedDecimals()` interface method.

File: packages/disputer/index.js (hunk 2 above)
Comment (chrismaree) [reply packages/disputer/index.js:112]: The whole point of these changes is to make it both forwards and backward compatible. to accommodate this the CLient needs to know: a) collateralTokenDecimals. b) syntheticTokenDecimals and c) priceFeedDecimals.

File: packages/disputer/test/index.js [hunk 6]
```diff
@@ -115,34 +115,45 @@ contract("index.js", function(accounts) {
...
-    collateralToken = await Token.new("DAI8", "DAI8", 8, { from: contractCreator });
-    constructorParams = {
-      ...constructorParams,
-      collateralAddress: collateralToken.address
-    };
-    emp = await ExpiringMultiParty.new(constructorParams);
+    collateralToken = await Token.new("BTC", "BTC", 8, { from: contractCreator });
+    syntheticToken = await SyntheticToken.new("Test Synthetic Token", "SYNTH", 18, { from: contractCreator });
+    // For this test we are using a lower decimal identifier, USDBTC. First we need to add it to the whitelist.
+    await identifierWhitelist.addSupportedIdentifier(padRight(utf8ToHex("USDBTC"), 64));
```
Comment (nicholaspai) [inline packages/disputer/test/index.js:128]: Should we refactor `padRight(someHex, 64)` into a common util file?

Comment (chrismaree) [reply packages/disputer/test/index.js:128]: ye, can do. should be in a separate PR though.

**Review by daywiss — APPROVED** [review]
looks good. It looks like this is mainly just updating the tests and there was very little that needed to be done in the actual disputer logic.

**Comment by chrismaree** [comment]
> looks good. It looks like this is mainly just updating the tests and there was very little that needed to be done in the actual disputer logic.

yup. The disputer does not compute any logic itself so min changes needed there.

**Review by nicholaspai — APPROVED** [review]
LGTM good tests

**Pull request merged**